torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,917 @@
+
+  // Outputs
+  insertUniquePredicated(
+      output_vals, consumer_edges, [](Val* v) { return !v->isFusionOutput(); });
+}
+
+std::ostream& operator<<(std::ostream& os, const SegmentedGroup* group) {
+  os << "g{";
+  for (size_t i = 0; i < group->exprs().size(); i++) {
+    os << group->exprs()[i]->name();
+    if (i + 1 != group->exprs().size())
+      os << ", ";
+  }
+  os << "}\n";
+  return os;
+}
+
+void SegmentedEdge::print() {
+  std::cout << "e{ \n  ";
+  from->print();
+  std::cout << " -> \n  ";
+  to->print();
+  std::cout << "through: ";
+  val->print();
+  std::cout << "}\\\\e\n\n";
+}
+
+std::ostream& operator<<(std::ostream& os, const SegmentedEdge* edge) {
+  os << "e{ " << edge->from << " -> " << edge->to << "(";
+  IrPrinter irp(os);
+  irp.handle(edge->val);
+  os << ") }\n";
+  return os;
+}
+
+SegmentedFusion::SegmentedFusion(const Fusion* fusion)
+    : fusion_(*fusion), impl_(this) {}
+
+void SegmentedFusion::print() {
+  std::cout << "Segmented_Fusion{ \n";
+  for (auto g : groups()) {
+    g->print();
+  }
+
+  for (auto e : edges()) {
+    std::cout << e << std::endl;
+  }
+  std::cout << "} //Segmented_Fusion\n";
+}
+
+SegmentedGroup* SegmentedFusion::Impl::makeGroup() {
+  groups_.emplace_back(std::make_unique<SegmentedGroup>());
+  return groups_.back().get();
+}
+
+SegmentedGroup* SegmentedFusion::Impl::makeGroup(Expr* expr) {
+  groups_.emplace_back(std::make_unique<SegmentedGroup>(expr));
+  return groups_.back().get();
+}
+
+SegmentedEdge* SegmentedFusion::Impl::makeEdge(
+    SegmentedGroup* from,
+    SegmentedGroup* to,
+    Val* val) {
+  edges_.emplace_back(std::make_unique<SegmentedEdge>(from, to, val));
+  return edges_.back().get();
+}
+
+void SegmentedFusion::Impl::cleanUnused() {
+  std::unordered_set<SegmentedGroup*> g_used(
+      owning_fusion_->groups().begin(), owning_fusion_->groups().end());
+  std::unordered_set<SegmentedEdge*> e_used(
+      owning_fusion_->edges().begin(), owning_fusion_->edges().end());
+
+  groups_.erase(
+      std::remove_if(
+          groups_.begin(),
+          groups_.end(),
+          [&g_used](auto& g) { return g_used.count(g.get()) == 0; }),
+      groups_.end());
+
+  edges_.erase(
+      std::remove_if(
+          edges_.begin(),
+          edges_.end(),
+          [&e_used](auto& e) { return e_used.count(e.get()) == 0; }),
+      edges_.end());
+}
+
+SegmentedGroup* SegmentedFusion::newGroup() {
+  SegmentedGroup* g = impl_.makeGroup();
+  groups_.push_back(g);
+  return g;
+}
+
+SegmentedGroup* SegmentedFusion::newGroup(Expr* expr) {
+  SegmentedGroup* g = impl_.makeGroup(expr);
+  groups_.push_back(g);
+  return g;
+}
+
+SegmentedEdge* SegmentedFusion::newEdge(
+    SegmentedGroup* from,
+    SegmentedGroup* to,
+    Val* val) {
+  SegmentedEdge* e = impl_.makeEdge(from, to, val);
+  edges_.push_back(e);
+  return e;
+}
+
+std::string SegmentedFusion::toString(int verbosity) const {
```

Comment:
please make this a namespace function instead (for consistency with kir::toString and other toString functions)

Reply:
Updated, Thanks!